Create postgresql_slot.py

jbscalia opened the pull request:
This is a new module for managing postgresql slots used in replication. I created this module as part of a replication role. Previously I had to use a shell script which was not idempotent.

##### SUMMARY
PostgreSQL uses slots for both physical and logical replication configurations, and prior to this module, no module performed management of slots.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
postgresql_slot

##### ANSIBLE VERSION
```
ansible 2.6.0 (postgresql_slot_branch e5b3416605) last updated 2018/05/07 11:10:54 (GMT -400)
  config file = /home/419635/.ansible.cfg
  configured module search path = [u'/home/419635/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/419635/ansible/lib/ansible
  executable location = /home/419635/ansible/bin/ansible
  python version = 2.7.5 (default, May  3 2017, 07:55:04) [GCC 4.8.5 20150623 (Red Hat 4.8.5-14)]
```

jbscalia committed:
Create postgresql_slot.py

This is a new module for managing postgresql slots used in replication. I created this module as part of a replication role. Previously I had to use a shell script which was not idempotent.

jbscalia committed:
Update postgresql_slot.py

Initial upload

jbscalia committed:
Update postgresql_slot.py

jbscalia committed:
Update postgresql_slot.py

jbscalia committed:
Update postgresql_slot.py

jbscalia committed:
Update postgresql_slot.py

jbscalia committed:
Update postgresql_slot.py

jbscalia committed:
Update postgresql_slot.py

Fixed the alias for name|slot and retested the module

jbscalia committed:
Update postgresql_slot.py

Removed typo

jbscalia committed:
Update postgresql_slot.py

Removed name and changed all references to use slot instead

jbscalia committed:
Update postgresql_slot.py

Missed removed the illegal option setting in the documentation section. Now fixed. Also removed the return at the end of main() as no other modules appear to have that construct.

jbscalia committed:
Update postgresql_slot.py

jbscalia committed:
Update postgresql_slot.py

Added missing slot: to RETURN seciton

jbscalia commented:
shipit

Andersson007 reviewed (changes requested):
@jbscalia , thank you for this good PR,

1. I think "short_description" should be swapped with "description".

2. I suggest to use "Adds or removes.." in descriptions (with "s" at the end of the words here like into other modules in PostgreSQL block), but it's not really important.

3. **db** - Add info that this is also the database to connect to. 

4.  "Name of the slot" Maybe shortly "Slot name"?

5. Maybe better to make state=absent and (decoder or type) as mutually exclusive?

6. Specify PostgreSQL version since Replication slots have been added in the "notes:" section.

7. Better to add a link to the official documentation about creating replication slots.

I want to try it if you make changes.

jbscalia commented:
Sure, I’ll take a look at these changes later today. Thanks for the feedback.

Sent from my iPad

> On Feb 1, 2019, at 3:54 AM, Andrey Klychkov <notifications@github.com> wrote:
> 
> @Andersson007 requested changes on this pull request.
> 
> @jbscalia , thank you for this good PR,
> 
> I think "short_description" should be swapped with "description".
> 
> I suggest to use "Adds or removes.." in descriptions (with "s" at the end of the words here like into other modules in PostgreSQL block), but it's not really important.
> 
> db - Add info that this is also the database to connect to.
> 
> "Name of the slot" Maybe shortly "Slot name"?
> 
> Maybe better to make state=absent and (decoder or type) as mutually exclusive?
> 
> Specify PostgreSQL version since Replication slots have been added in the "notes:" section.
> 
> Better to add a link to the official documentation about creating replication slots.
> 
> I want to try it if you make changes.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

jbscalia committed:
Update postgresql_slot.py

modified per Andrey's suggestions, I hope everything is acceptable now.

jbscalia committed:
Update postgresql_slot.py

jbscalia committed:
Trying to satisfy the Ansibot

Missed changing the return value

jbscalia commented:
my version runs correctly on my server, so I', hoping the Ansibot finds nothing wrong, but it seems awfully fussy.

Andersson007 commented on lib/ansible/modules/database/postgresql/postgresql_slot.py:
```diff
@@ -0,0 +1,227 @@
+slot_name:
+    description: Name of the slot
+    returned: success, changed
+    type: string
```
change 'string' to 'str'
this recent change in Ansible ('str', 'bool', etc.)

Andersson007 replied:
and remove a blank lines near line 226

jbscalia committed:
Changes suggested have been completed

I changed "string" to read "str" as you noted, and removed several duplicate blank lines between functions and sections, each now only has a single blank line between them. I hope this is what you meant in your suggestion.

jbscalia committed:
Fixed latest Ansibot complaints

Removed errant colon on line 155 and added several 2nd blank lines after Ansibot complaints.

jbscalia committed:
Latest complaints fixed

The Ansibot complained about several missing blank second lines which I removed in the belief that I was instructed to remove them. We'll see what it complains about now.

jbscalia committed:
More typos

The editor added a extra blank line at the end and it does a very poor job when including a new blank line as it might add spaces to that new line. This would cause more errors.

jbscalia committed:
Final typo, I hope

The if statement on line 223 had the terminating colon missing, not sure how that happened

Andersson007 commented:
don't panic ) just 10 minutes

Andersson007 commented:
Did you use this check in your repo dir?
```
ansible# source ./hacking/env-setup
ansible# ansible-test sanity --python 2.7  --base-branch devel lib/ansible/modules/database/postgresql/postgresql_slot.py
```
I'm going to send you proper file in 10 minutes.

Andersson007 commented:
@jbscalia , I just sent fixed file to your mail, please, swap it and push.
On Monday I test this functionality.

jbscalia commented:
I received your file, but I can only access it from work on my iPhone. The blasted government will not permit me to connect to gmail here. Will swap out the old file for yours once I get home.

Sent from my iPhone

> On Feb 1, 2019, at 1:11 PM, Andrey Klychkov <notifications@github.com> wrote:
> 
> @jbscalia , I just sent fixed file to your mail, please, swap it and push.
> On Monday I test this functionality.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

kostiantyn-nemchenko commented on lib/ansible/modules/database/postgresql/postgresql_slot.py:
```diff
@@ -0,0 +1,226 @@
+    description:
+      - Name of the database where you're connecting in order to add or remove only the logical slot to/from
+    required: false
+  decoder:
```
Would it be better to stick with the naming convention used in Postgres [docs](https://www.postgresql.org/docs/current/view-pg-replication-slots.html) and use `plugin` or `output_plugin` instead of `decoder`

kostiantyn-nemchenko commented on lib/ansible/modules/database/postgresql/postgresql_slot.py:
```diff
@@ -0,0 +1,226 @@
+
+    db = module.params["db"]
+    slot = module.params["slot_name"]
+    type = module.params["type"]
```
It's generally not a good idea to mask [built-in](https://docs.python.org/2/library/functions.html#type) functions.
One more reason to pick `slot_type` as per the [docs](https://www.postgresql.org/docs/current/view-pg-replication-slots.html).

kostiantyn-nemchenko commented on lib/ansible/modules/database/postgresql/postgresql_slot.py:
```diff
@@ -0,0 +1,226 @@
+            db=dict(required=False),
+            slot_name=dict(required=True),
+            type=dict(default="physical", choices=["physical", "logical"]),
+            decoder=dict(default="test_decoding"),
```
Since `physical` is default I suggest leaving this empty

kostiantyn-nemchenko commented on lib/ansible/modules/database/postgresql/postgresql_slot.py:
```diff
@@ -0,0 +1,226 @@
+notes:
+  - The default authentication assumes that you are either logging in as or sudo'ing to the C(postgres) account on the host.
+  - Physical replication slots were introduced to PostgreSQL with version 9.4, while logical replication slots were added beginning with version 10.0.
+    Physical slots are explained in https://www.postgresql.org/docs/9.4/protocol-replication.html and logical slots which were a later edition are
```
Hardcoded major versions should be replaced by "current" to always provide up to date pages for end users

dagwieers commented:
@jbscalia What exactly is it you are trying to do ? If you make changes to commits you already pushed, then git will complain and the only thing you can do is to force-push those changes. But I am not sure this is what you wanted to do.

Also you can save yourself the time going back-and-forth with CI testing byt running `ansible-test sanity <path/to/your/module>` and fix everything before pushing it to GitHub. We are doing automated testing that goes beyond simply running it.

jbscalia commented:
Dag,

I was trying to replace the file with a corrected version given to me by Andresson, and I had logged in and tried to set edit. I was pretty certain that the replacement file would pass all tests, but github kept crashing when I hit edit. Now, none of this was happening when I was editing this at my office, only here at home, and since I’d been working on this for about 10 hours, I was just more than a bit frustrated.

Sent from my iPad

> On Feb 3, 2019, at 5:47 PM, Dag Wieers <notifications@github.com> wrote:
> 
> @jbscalia What exactly is it you are trying to do ? If you make changes to commits you already pushed, then git will complain and the only thing you can do is to force-push those changes. But I am not sure this is what you wanted to do.
> 
> Also you can save yourself the time going back-and-forth with CI testing byt running ansible-test sanity <path/to/your/module> and fix everything before pushing it to GitHub. We are doing automated testing that goes beyond simply running it.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

dagwieers commented:
@jbscalia I think it is easier to learn to use git command-line (or using an editor with a Git interface). Being able to use `ansible-test` is pretty much a requirement unless you fancy getting corrected by @ansibot ;-)

PS You can also give @Andersson007 write-access to your branch so he can continue/finish with this PR.

jbscalia commented:
Hmm, i’ve never tried using a command line for github, although I’m quite familiar with git. I’ll look at it.

Sent from my iPad

> On Feb 3, 2019, at 6:54 PM, Dag Wieers <notifications@github.com> wrote:
> 
> @jbscalia I think it is easier to learn to use git command-line (or using an editor with a Git interface). Being able to use ansible-test is pretty much a requirement unless you fancy getting corrected by @ansibot ;-)
> 
> PS You can also give @Andersson007 write-access to your branch so he can continue/finish with this PR.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

jbscalia commented:
I've checked and do not have approval for downloading this repo so I could do the command line operations. Sigh... How would I go about granting Andersson007 write access so he could do the adjustments?

dagwieers commented:
@jbscalia It is described here: https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/

It basically is adding him as a collaborator to your forked repository.

jbscalia committed:
Fixed latest round of ansibot's complaints

Went through the last message from the Ansibot and corrected the extra space issues it was complaining about. Still not exactly certain about the final line error it complained about. We'll see.

jbscalia commented:
It does appear to me like Andressen007 and other maintainers can make changes to this pull request. At least I'm seeing the checkbox is in fact, checked.

jbscalia committed:
Removed blank lines at end of file

We'll see if the editor puts these back in, but this edit should finally satisfy the ansibot.

jbscalia committed:
One last change, hopefully

It appears that I needed one more blank line on line 223. At least I think that's what ansibot complained about.

jbscalia committed:
updated type to be slot_type

jbscalia committed:
changed decoder to output_plugin

jbscalia committed:
missed one correction for output_plugin.

Now changed.

kostiantyn-nemchenko reviewed (changes requested):
Sorry, didn't notice that earlier, but could you please add support of using a [unix socket](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/database/postgresql/postgresql_ext.py#L45), [ssl-related options](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/database/postgresql/postgresql_ext.py#L49) and recently added [session role](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/database/postgresql/postgresql_ext.py#L69)?

jbscalia commented:
> We also need integration tests (You should write it).

Are there some examples of these tests and where should I place them? This is a fairly new experience for me.

Andersson007 commented:
@jbscalia you can look at examples of postgresql integration tests here ansible/test/integration/targets/postgresql/defaults/

https://github.com/ansible/community/issues/435 - There are some useful (and necessary) information from @dagwieers about integration tests.
See also official documentation:
https://docs.ansible.com/ansible/latest/dev_guide/testing_integration.html

If you're using a linux machine for tests (you should do it), you can test it before pushing your commit:
1. Prepare your system:
```
# pip install argcomplete
# pip3 install argcomplete
# yum install docker
# systemctl enable docker && systemctl start docker
```
From ansible src root directory:
```
# source ./hacking/env-setup
```

2. Run test in the docker container:
```
test/runner/ansible-test integration --docker-no-pull --docker fedora28 -v postgresql >> ~/test.log
```
3. Parse log file for tests result, e.g.
```
less ~/test.log
```
Use searching in less, e.g.:
```
/postgresql_slot
```
Maybe there is easier way, I don't know, but seems this method works well.

jbscalia commented:
> Sorry, didn't notice that earlier, but could you please add support of using a [unix socket](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/database/postgresql/postgresql_ext.py#L45), [ssl-related options](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/database/postgresql/postgresql_ext.py#L49) and recently added [session role](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/database/postgresql/postgresql_ext.py#L69)?

I'm not seeing anything the current PostgreSQL documents regarding what you've suggested for this slot module. I am seeing a new parameter though that this does not cover: immediately_reserve.  So, I guess I should be adding that.

kostiantyn-nemchenko commented:
> I'm not seeing anything the current PostgreSQL documents regarding what you've suggested for this slot module.

@jbscalia Have you checked the links I posted above?
I mean `login_unix_socket`, `ssl_mode`, `ssl_rootcert` and `session_role` options which should be present in all Postgres-related modules for Ansible.

jbscalia commented:
>> I mean `login_unix_socket`, `ssl_mode`, `ssl_rootcert` and `session_role` options which should be present in all Postgres-related modules for Ansible.

OK, I see what you mean, sort of. I've looked at some other postgresql modules and I see everything except session_role. Is that parameter new with Ansible 2.8? It's not present in any other module on my host server.

kostiantyn-nemchenko commented:
> Is that parameter new with Ansible 2.8?

Yes, it has been added recently by https://github.com/ansible/ansible/pull/43650 and will be in the next release.

jbscalia commented:
I've gotten the code adjusted on my server so it includes Kostiantyn's parameters and I've added both immediately_reserve as well as temporary, but I'm wondering if temporary is really necessary. Mind you it appears that I must make the creation call with either 1 argument or all 3 arguments, 2 arguments does not work, but since temporary says the slot will only exist for this session, I think it would disappear after the play is completed, which wouldn't be of any real use.

BTW, I'm having some problems getting the integration tests running as the docker installation is failing due to a dependency check with yum. Still trying to work that out.

jbscalia commented:
Yes, I was correct about the temporary parameter. Ran a test with that set to True and there is no slot on the server afterwards. So, unless someone objects vociferously I'm going to hard code that to False and not allow it be settable.

jbscalia committed:
Update postgresql_slot.py

jbscalia committed:
line 199 was missing an equals sign

And I changed the choices to be double-quoted like the other ones are shown.

jbscalia committed:
Fixed some spacing issues

Once again, the editor added extra spaces to some blank lines

jctanner commented:
!waffling ci_verified

jbscalia committed:
Another attempt 

Although there are a couple of errors I actually do not understand, but I'm being rushed.

jbscalia committed:
I think this correct now

But, I don't really know why the ssl_mode choices are saying they don't match with the documentation. I'm not seeing anything different.

jbscalia committed:
Trying to satisfy the ansibot

Reworked how the definitions for those parameters which are not required, but I still think these are incorrect.

jbscalia committed:
Reverted immediate_reserve back to a boolean

Last edit obviously didn't work, resetting this back to a boolean

kostiantyn-nemchenko commented:
@jbscalia Please, take a look at [this](https://github.com/jbscalia/ansible/pull/1) PR.

Andersson007 commented:
@jbscalia hello,
are you going to fix this module? If yes, cool, I'll make a full review after that.
If you don't have enough time or for other reasons, you don't mind if I (or anybody else) completely rewrite this module (keeping you as an co-author, of course) ?

Andersson007 commented:
!needs_info

Akasurde commented:
superseded by #55079

Andersson007 commented:
@Akasurde  , thank you
